compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -671,6 +676,8 @@ public void visit(BLangQueryAction queryAction) {
             }
         }
         analyzeReachability(queryAction.getDoClause(), env);
+        queryAction.returnsWithinDoClause = this.returnedWithinQuery;
+        this.returnedWithinQuery = prevReturnedWithinQuery;
```

Comment:
The behavior for the following program is not correct. Since the condition in the `where` clause will never be satisfied, it doesn't execute the statements in the `do` clause.

```ballerina
public function main() {
    string a = testReachabilityWithQueryAction();
    io:println(a.toBalString()); // prints (), which is wrong
}

function testReachabilityWithQueryAction() returns string {
    string?[] stringArray = [(), (), ()];

    error? unionResult = from var item in stringArray
        where item is string
        do {
                return "Hello";
        };
}
```

I think the function should have a return value if it doesn't return from the query action.
Due to https://github.com/ballerina-platform/ballerina-lang/issues/34914, we are unable to include a statement after the query action ATM. Adding https://github.com/ballerina-platform/ballerina-lang/blob/70ebe0903df9f539bf505c39185a2f1a77f3710b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java#L701-L702 in the visitor of the `DoClause` would solve the issue (The fix will be added in https://github.com/ballerina-platform/ballerina-lang/pull/35287 ).

And even if we fix this and add a return statement at the end as given below, the result would still be the same.
```ballerina
public function main() {
    string a = testReachabilityWithQueryAction();
    io:println(a.toBalString()); // prints (), which is wrong. should print "World"
}

function testReachabilityWithQueryAction() returns string {
    string?[] stringArray = [(), (), ()];

    error? unionResult = from var item in stringArray
        where item is string
        do {
                return "Hello";
        };

    return "World";
}
```

Reply:
Improved with b6e47c0acab3b4f681a06a755a855c8b833c0d9e to handle this. But facing https://github.com/ballerina-platform/ballerina-lang/issues/35383. We'll have to rethink how to handle such a scenario because we consider nil as the successful completion type to continue the loop. Hence created this separate issue https://github.com/ballerina-platform/ballerina-lang/issues/35383